src/componmnet/PointCloud/PointCloud.jsx

```diff
@@ -1,61 +1,193 @@
+        setLegendDisplay(true);
+        const newDivs = [];
+        const pool = is3D ? ref3D.nodePool.current : ref2D.nodePool.current;
+        const colorPool = generateCategoryColor(colorData, pool);
+        for (const [key, value] of Object.entries(colorData.cat_mapper)) {
+            const colorStr = colorPool[key].substr(2);
+            const r = parseInt(colorStr.substr(0, 2), 16);
+            const g = parseInt(colorStr.substr(2, 2), 16);
+            const b = parseInt(colorStr.substr(4, 2), 16);
+
+            newDivs.push(
+                <div key={value} style={{margin: 0, display: "flex", 
+                                        flexDirection: "row", justifyContent:'flex-end'}}>
+                    <div style={{backgroundColor: `rgb(${r}, ${g}, ${b})`, width: '15px', height: '15px'}}></div>
+                    <p style={{paddingLeft: '10px'}}>{value}</p>
                 </div>
+            );
+        }
+        setLegend(newDivs);
+    }
 
-                <div className='zoomPanel'>
-                    <PlusSquareFilled />
-                    <MinusSquareFilled />
+    const resetColor = () => {
+        setLegendDisplay(false);
+        const pool = is3D ? ref3D.nodePool.current : ref2D.nodePool.current;
+        generateRandomColor(pool);  
+    }
+
+    const inCavasPanel = (setUI) => (
+        <div>
+            <div className='buttonPanel'>
+                <div className='buttonLeftGroup'>
+                    <CameraOutlined className='downloadIcon'></CameraOutlined>
+                    <Button onClick={() => setUI(0)}>Point Cloud</Button>
+                    <Button onClick={() => setUI(1)}>Network</Button>
+                </div>
+                <div className='buttonRightgroup'>
+                    <Button onClick={gradientColor}>Color Gradient</Button>
+                    <Button onClick={categoryColor}>Color Group</Button>
+                    <Button onClick={resetColor}>Reset</Button>
+                    <div style={{paddingTop: '20px', display: "flex", 
+                                flexDirection: "column", alignItem: 'flex-end', 
+                                display: legendDisplay ? '' : 'none'}}>
+                        {legend}
+                    </div>
+                    
                 </div>
             </div>
+
+            <div className='zoomPanel'>
+                <PlusSquareFilled />
+                <MinusSquareFilled />
+            </div>
         </div>
+    )
+
+    return (
+        <div className='pointCloud'>
+            <Panel  is3D={is3D} setIs3D={setIs3D}
+                    setLegendDisplay={setLegendDisplay} 
+                    width={window.innerWidth * style.widthRatio}>    
+            </Panel>
+            <div style={{position: 'relative'}}>
+                {/* the map way is working but it did not seem to shorten the code and added confusion */}
+                {
+                    [ref2D, ref3D].map(it => {
```

Comment:
```suggestion
                    [ref2D, ref3D].map((it, index) =>
```

This way you will have key